### ISSUE TITLE
[Question] How to Add/Update Hazelcast config through the operator/ Hazelcast CRD

### ISSUE THREAD

**Issue by bokjo** [hazelcast/hazelcast-platform-operator#384]
Hello all,

Maybe not the right place to ask this one.

Is there a way to add/update the Hazelcast config through the Hazelcast CRD somehow, I'm seeing just predefined options in the spec that are available.

For example disabling `Jet` in the CRD, since it is enabled by default?

Editing the ConfigMap after the cluster is created is picked up by the operator controller but nothing is happening (reconciling on StatefulSet side)... manually restarting the cluster StatefulSet doesn't pick up the manually updated ConfigMap.

```json
{"level":"info","ts":1665067430.970415,"logger":"controllers.Hazelcast","msg":"Operation result","hazelcast":"hazelcast/hz","ConfigMap":"hz","result":"updated"}
```

Seems there is no way to set the `near-cache` options in the `Map` CRD, is this by design or there is some other way to do it?

Also, I'm using tools like `crd2pulumi` to generate code from the CRDs and is was failing because there were default values for object types  like 

```yaml
          spec:
            # default: # NOTE: Che hack - removed because crd2pulumi was failing
            #   repository: docker.io/hazelcast/hazelcast
            description: HazelcastSpec defines the desired state of Hazelcast
            properties:
              agent:
                # default: # NOTE: Che hack - removed because crd2pulumi was failing
                #   repository: docker.io/hazelcast/platform-operator-agent
                #   version: 0.1.6
                description: B&R Agent configurations
```
is this a good place to report those, think they are not actually needed!

Thanks in advance
@bokjo

**Comment by hasancelik** [comment]
Hi @bokjo,

> Maybe not the right place to ask this one.

This repo is the best place to ask these questions 👍 🙂 

> Is there a way to add/update the Hazelcast config through the Hazelcast CRD somehow, I'm seeing just predefined options in the spec that are available.

The whole idea of the platform operator is to make an abstraction over the Hazelcast configuration to ease the DevOps processes. We are adding multiple data structures and config options with every release. At some point, the Platform Operator will cover almost every config scenario. BTW, we already have [helm charts ](https://github.com/hazelcast/charts), which have full control over the Hazelcast configuration.

> For example disabling Jet in the CRD, since it is enabled by default?

Good point, we can add a config option for that 👍 

> Editing the ConfigMap after the cluster is created is picked up by the operator controller but nothing is happening (reconciling on StatefulSet side)... manually restarting the cluster StatefulSet doesn't pick up the manually updated ConfigMap.

This usage is not supported unfortunately, the configmap is overwritten by Hazelcast CR reconcile logic.

> Seems there is no way to set the near-cache options in the Map CRD, is this by design or there is some other way to do it?

Near-cache support for Map CR in out roadmap 👍 

> Also, I'm using tools like crd2pulumi to generate code from the CRDs and is was failing because there were default values for object types like

Why do you need that?

> is this a good place to report those, think they are not actually needed!

As I said above, every feedback or feature request is very valuable to us.

**Comment by bokjo** [comment]
> 

Hello @hasancelik 

Thanks for taking a look .

> Also, I'm using tools like crd2pulumi to generate code from the CRDs and is was failing because there were default values for object types like
> 
> Why do you need that?

We are using Pulumi for deploying everything to K8s depending on an env./cluster (not applying YAML configs directly). I'm converting the CRD YAML to Typescript code and using it accordingly afterward to deploy things.

basically, the operator deployment for me is something like this

```ts
const hzPlatformOperator = new HazelcastPlatformOperator(...);
```
It is not a showstopper, for now, we will just need to comment out those lines in the CRDs whenever a new version comes out, no idea whether those are actually needed or used for something in the CRD specification (the individual fields already have default values)

Some other remarks

- it will be nice to be able to enable the `Memcached` mode through the Hazelcast CRD as well (no option for it)

- pp.s is there a public roadmap for the operator features and their cadence, at least some place to see and suggest some if needed?

p.s I'm aware of the Helm chart, but not really a fan of Helm in general :)

**Comment by bokjo** [comment]
Hello @hasancelik ,

Any update/plan about this one?

**Comment by hasancelik** [comment]
Hi @bokjo,

`default` values are `must have` for our operator logic; unfortunately, we can not remove them from CRD definition YAML. These values are checked by the reconciliation logic, and take action based on them. It seems this issue should be resolved at crd2pulumi repo; I already see multiple issues at the repo:
https://github.com/pulumi/crd2pulumi/issues/22
https://github.com/pulumi/pulumi/blob/0f9273fc7bba786dcc1e1e85008f3fabe9cfabec/pkg/codegen/schema/schema.go#L1322

> It will be nice to be able to enable the Memcached mode through the Hazelcast CRD as well (no option for it) 

Do you mean Memcache client support? 
https://docs.hazelcast.com/hazelcast/5.2/clients/memcache#memcache-client-code-examples

> Is there a public roadmap for the operator features and their cadence, at least some place to see and suggest some if needed?

We do not have it for now, unfortunately, but I will discuss this idea with the team 👍 

> I'm aware of the Helm chart, but not really a fan of Helm in general
🙂 

BTW, some updates about changes you requested at first comment:

> For example disabling Jet in the CRD, since it is enabled by default?

This is already available with 5.5 release:
https://github.com/hazelcast/hazelcast-platform-operator/blob/main/config/samples/_v1alpha1_hazelcast_jet.yaml#L8

> Seems there is no way to set the near-cache options in the Map CRD, is this by design or there is some other way to do it?

We have added it to roadmap, probably it will be available with 5.7 release, I think.
